<a href="https://colab.research.google.com/github/jppeirce/DSC210-Foundations-of-Data-Science/blob/main/Homework/HW02-python_and_arrays.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 2: Python Foundations and NumPy Arrays

**DSC 210 Foundations of Data Science**

*Covers Module 3 (Introduction to Python) and Module 4, Section 1 (arrays, broadcasting, and boolean masks).*

References:
- Module 3 and Module 4 course notebooks
- [Hands-on Introduction to Data Science with Python](https://florian-huber.github.io/data_science_course/) (CC BY-NC-SA 4.0)
- `penguins` and `titanic` datasets, included with the `seaborn` package

*Last major revision: 2026-07-30*

## Key Concepts

- Predict the type and value of Python expressions before running them
- Index, slice, and safely copy lists
- Explain why lists fail for element-wise math and arrays succeed
- Use broadcasting and boolean masks to ask questions of real data
- Choose the right denominator, because a fraction is only as honest as what it is out of

## Instructions

**Total: 50 points.** Point values are listed with each problem.

- Answer written questions in the markdown cells marked **Your answer:**.
- Cells marked `# FILL-IN` contain blanks written as `____` and will raise a `SyntaxError` until every blank is replaced. That error is expected until you finish the problem.
- Several problems ask you to **predict before you run**. Write the prediction down first. A wrong prediction costs nothing; a missing one does.
- Unlike Homework 1, this assignment asks you to **write and modify code yourself**. Everything you need appeared in Modules 3 and 4.
- Submit the completed `.ipynb` file, with all cells run, to Canvas.

---
## Part 0. Setup check (2 points)
---

Run the cell below. It loads the packages and both datasets used in this assignment.

In [1]:
import seaborn as sns
import numpy as np

penguins = sns.load_dataset('penguins')
titanic  = sns.load_dataset('titanic')

print('numpy version :', np.__version__)
print('penguins shape:', penguins.shape)
print('titanic shape :', titanic.shape)

numpy version : 2.1.3
penguins shape: (344, 7)
titanic shape : (891, 15)


**Problem 0.** (2 points) Did the cell run without an error? If yes, report the two shapes it printed. If no, paste the error message.

**Your answer:** Yes, penguins shape: (344, 7), titanic shape : (891, 15)



---
## Part A. Short answer (12 points)
---

In [74]:
print(type(15 / 3) #float	5.0
,type(15 * 4)	#int	60
,type('DSC' + '210')	#string	'DSC210'
#type('DSC' + 210)
,type(3 > 5)	#boolean	FALSE
,type(True + True + False) #int	2
,15 / 3 #float	5.0
,15 * 4	#int	60
,'DSC' + '210'	#string	'DSC210'
#'DSC' + 210
,(3 > 5)	#boolean	FALSE
,True + True + False) #int	2

<class 'float'> <class 'int'> <class 'str'> <class 'bool'> <class 'int'> 5.0 60 DSC210 False 2


### Problem A1. Predict type and value (3 points)

For each expression, state the **type** of the result and the **value**, or state that it raises an error. Do this *without* running code first; you may check afterward and note any you got wrong.

| # | Expression | Type | Value (or error) |
| --- | --- | --- | --- |
| a | `15 / 3` | float | 5.0 |
| b | `15 * 4` | int | 60 |
| c | `'DSC' + '210'` | string | 'DSC210' |
| d | `'DSC' + 210` | error | error |
| e | `(3 > 5)` | boolean | FALSE |
| f | `True + True + False` | int | 2 |

**Your answer:** *(fill in the table above, and note below which ones surprised you when you checked)*
None suprised me.


### Problem A2. Indexing and slicing (3 points)

The list below includes the daily closing price of a stock for one trading week, Monday through Friday:

```python
prices = [102.5, 101.0, 104.2, 103.8, 106.1]
```

(a) Write the expression that returns Wednesday's price.

(b) Write the expression that returns Thursday and Friday as a list.

(c) A classmate writes `prices[1:2]` expecting Monday and Tuesday. They have made **two** distinct mistakes. Identify both.

**Your answer:**

(a) prices[2]

(b) prices[3:]

(c) Mistake 1: Indicies in python start at zero so the first input should be 1. Mistake 2: When using ':' python selects all values from the index before the ':' to the index after the ':' not including the value at that index. The provided code will return a list of length 1.


### Problem A3. The silent bug (3 points)

Consider this code, which runs without any error:

```python
raw_data = [4.2, 5.1, 3.9, 4.8]
cleaned = raw_data
cleaned.append(0.0)
```

(a) After the third line, what does `raw_data` contain? Explain why in one sentence.

(b) Rewrite the second line so that `raw_data` is protected.
**Your answer:**

(a) raw_data contains [4.2, 5.1, 3.9, 4.8, 0.0] because the cleaned varriable points to raw_data list meaning they are the same variable.

(b) cleaned = raw_data.copy


### Problem A4. Lists versus arrays (3 points)

(a) Predict the output of each line:

```python
[70, 80, 90] + [70, 80, 90]
np.array([70, 80, 90]) + np.array([70, 80, 90])
```

(b) A NumPy array can hold only one type. Predict what `np.array([True, 2, 3.5])` contains, and what `np.array([1, 2, 'three'])` contains.

**Your answer:**

(a) [70, 80, 90, 70, 80, 90] and
    [140, 160, 180]

(b) [1.0, 2.0, 3.5] and ['1', '2', 'three']

---
## Part B. Code completion (12 points)
---
Each problem is one `# FILL-IN` cell. Replace every `____` with code (not the asnwer), run the cell, and make sure the output makes sense before moving on.

### Problem B1. Summarizing a list (4 points)

A field team measured the body masses, in grams, of eight penguins.

In [37]:
# FILL-IN  (Problem B1)
masses = [3750, 3800, 3250, 3700, 3450, 3650, 3800, 5700]

# (a) the number of penguins in the sample
print('count   :', len(masses))

# (b) the lightest and heaviest
print('lightest:', min(masses))
print('heaviest:', max(masses))

# (c) the mean mass, rounded to 1 decimal place
print('mean    :', round(sum(masses) / len(masses), 1))

# (d) the first three penguins, as a list
print('first 3 :', masses[0:3])

count   : 8
lightest: 3250
heaviest: 5700
mean    : 3887.5
first 3 : [3750, 3800, 3250]


One penguin is far heavier than the rest. In one sentence: does removing it change the mean or the median more, and why?

**Your answer:** Removing the upper outlier penguin would change the mean more. This is because the mean is sensitive to outliers--unlike the median.



### Problem B2. Broadcasting (4 points)

The cell below pulls flipper lengths from the penguins dataset as a NumPy array.

In [41]:
# FILL-IN  (Problem B2)
flippers_mm = penguins['flipper_length_mm'].dropna().to_numpy()

# (a) convert every measurement from millimeters to centimeters,
# using broadcasting
flippers_cm = flippers_mm.copy() / 10
print('first five (cm):', flippers_cm[:5])

# (b) how many flipper measurements are there?
print('count:', len(flippers_mm))

# (c) mean flipper length in cm, rounded to 2 decimal places
print('mean (cm):', round(sum(flippers_cm) / len(flippers_cm), 2))

first five (cm): [18.1 18.6 19.5 19.3 19. ]
count: 342
mean (cm): 20.09


### Problem B3. Boolean masks (4 points)

In [61]:
# FILL-IN  (Problem B3)
mass_g = penguins['body_mass_g'].dropna().to_numpy()

# (a) a boolean mask for penguins over 4000 g
big = mass_g > 4000
print('first five of the mask:', big[:5])

# (b) how many penguins are over 4000 g?
print('count over 4000 g:', sum(big))

# (c) the fraction of all penguins over 4000 g
#     (recall: the mean of a boolean array is a proportion)
print('fraction:', round(big.mean(), 3))

# (d) the mean mass of ONLY the penguins over 4000 g
print('mean of the big ones:', round(mass_g[big].mean(), 1))

first five of the mask: [False False False False False]
count over 4000 g: 172
fraction: 0.503
mean of the big ones: 4863.4


Parts (c) and (d) both use the mask, but they answer different questions. State each question in plain English.

**Your answer:** (c) What proportion of penguins are big (d) What is the average mass of big penguins.



---
## Part C. Guided coding (12 points)
---
Now you write the code. Every tool you need appeared in Modules 3 and 4.

### Problem C1. Children aboard the Titanic (6 points)

Homework 1 asked you to *read* the titanic data. Now you can *query* it.

Write code in the cell below to answer all four parts. Use comments to label which part each line answers.

(a) Pull the `age` column into a NumPy array called `ages`, dropping missing values first (the Module 4 pattern: `.dropna().to_numpy()`).

(b) Report how many passengers have a recorded age, and the mean age rounded to 1 decimal place.

(c) Using a boolean mask, report how many passengers were children (under 18), and what fraction of passengers with a recorded age that represents.

(d) Report the mean age of the children only.

In [64]:
# Problem C1: your code here
# (a) Pull the age column into a NumPy array called ages, dropping missing values first (the Module 4 pattern: .dropna().to_numpy()).
ages = titanic['age'].dropna().to_numpy()
# (b) Report how many passengers have a recorded age, and the mean age rounded to 1 decimal place.
print('count:', len(ages))
print('mean:', round(sum(ages) / len(ages), 1))
# (c) Using a boolean mask, report how many passengers were children (under 18), and what fraction of passengers with a recorded age that represents.
children = ages < 18
print('count of children:', sum(children))
print('fraction of children:', round(children.mean(), 3))
# (d) Report the mean age of the children only.
print('mean of children:', round(ages[children].mean(), 1))


count: 714
mean: 29.7
count of children: 113
fraction of children: 0.158
mean of children: 9.0


(e) In Homework 1 you found that about 177 passengers have no recorded age. Your fraction in (c) is therefore a fraction *of passengers with a recorded age*, not of everyone aboard. Give one plausible reason the missing ages might not be spread evenly across passengers, and say which direction that could bias your estimate of the fraction of children.

**Your answer:** This may be because childrens ages may be more stricly documented compared to adults. The fraction of children may be smaller because this suggestion makes it more likely that more the 177 unrecorded age passengers are adults. This means that our estimate is biased up and is a greater value than it should be.



### Problem C2. Playing with Penguns (6 points)

Are Gentoo penguins heavier than Adelie penguins? Answer with arrays and masks.

(a) Build a NumPy array of body masses and a matching array of species names. Start from `penguins.dropna()` so the two arrays line up row for row.

(b) Build boolean masks `adelie` and `gentoo` from the species array.

(c) Report the count and mean mass for each species, with the means rounded to 1 decimal place.

(d) Report the difference between the two means.

In [73]:
# Problem C2: your code here
# (a) Build a NumPy array of body masses and a matching array of species names. Start from penguins.dropna() so the two arrays line up row for row.
cleaned = penguins.dropna()
mass_g = cleaned['body_mass_g'].to_numpy()
species = cleaned['species'].to_numpy()
# (b) Build boolean masks adelie and gentoo from the species array.
adelie = (species =='Adelie')
gentoo = (species =='Gentoo')
# (c) Report the count and mean mass for each species, with the means rounded to 1 decimal place.
print('Adelie count:',sum(adelie))
print('Gentoo count:',sum(gentoo))
print('Adelie mean:', round(mass_g[adelie].mean(), 1))
print('Gentoo mean:', round(mass_g[gentoo].mean(), 1))
# (d) Report the difference between the two means.
print('Difference between Adelie versus Gentoo:', round(mass_g[adelie].mean() - mass_g[gentoo].mean(), 1))

Adelie count: 146
Gentoo count: 119
Adelie mean: 3706.2
Gentoo mean: 5092.4
Difference between Adelie versus Gentoo: -1386.3


(e) Write your conclusion as one sentence a non-programmer could read. Then, in a second sentence, name one thing your comparison does **not** establish. (Note: Module 4's warning about pooled comparisons)

**Your answer:** Conclusion: Gentoo penguins are heavier than Adelie peguins. This conclusion does not establish if the island these values are sampled from has an effect on the the difference between Adelie versus Gentoo.



---
## Part D. Application (12 points)
---

### Problem D1. Who survived? (6 points)

The `titanic` dataset records `survived` as 0 or 1, and `sex` as `'male'` or `'female'`. Because the mean of a 0/1 array is a proportion, boolean masks let you compute survival rates directly.

(a) (3 points) Write code to report three numbers: the overall survival rate, the survival rate among women, and the survival rate among men. Round each to 3 decimal places. (Pull the columns with `.to_numpy()`; there are no missing values in these two columns.)

In [87]:
# Problem D1(a): your code here
#(a) (3 points) Write code to report three numbers: the overall survival rate, the survival rate among women, and the survival rate among men.
#Round each to 3 decimal places. (Pull the columns with .to_numpy(); there are no missing values in these two columns.)
survived = titanic['survived'].to_numpy()
print('Survival rate:', round(survived.mean(), 3))
sex = titanic['sex'].to_numpy()
print('Survival rate among women:', round(survived[sex == 'female'].mean(), 3))
print('Survival rate among men:', round(survived[sex == 'male'].mean(), 3))
#(b) (1 point) Also compute: among the survivors, what fraction were women? (Mask to survivors first, then ask about sex.)
survivors_sex = sex[survived == 1]
print('Fraction of surviving women:', round((survivors_sex == 'female').mean(), 3))

Survival rate: 0.384
Survival rate among women: 0.742
Survival rate among men: 0.189
Fraction of surviving women: 0.681


(b) (1 point) Also compute: *among the survivors*, what fraction were women? (Mask to survivors first, then ask about sex.)

(c) (2 points) The number in (b) and the survival rate among women are different questions with different denominators. A journalist writes: "Most survivors were women, so being a woman nearly guaranteed survival." Using your numbers, explain what the journalist got right and what they got wrong. Name the denominator of each claim.

**Your answer:**

(b) 68.1% of survivors

(c) The journalist was right that most survivors were Women. However, being a woman does not guarantee survival because the proportion of people who survived is not taken into account. Of women, the proportion that survived is much smaller than the proportion of survivors who are women.


### Problem D2. Which mean answers the question? (6 points)

A weather station records daily rainfall, in millimeters, for three weeks:

```python
rain = np.array([0, 0, 3.2, 11.4, 0, 0.6, 0,
                 0, 0, 22.1, 4.5, 0, 0, 1.1,
                 0, 8.3, 0, 0, 2.7, 0, 0])
```

(a) (3 points) Write code to compute: the number of rainy days (rainfall greater than 0), the fraction of days that were rainy, the mean rainfall over **all** days, and the mean rainfall over **rainy days only**. Round the means to 2 decimal places.

In [89]:
# Problem D2(a): your code here
rain = np.array([0, 0, 3.2, 11.4, 0, 0.6, 0,
                 0, 0, 22.1, 4.5, 0, 0, 1.1,
                 0, 8.3, 0, 0, 2.7, 0, 0])
#(a) (3 points) Write code to compute: the number of rainy days (rainfall greater than 0), the fraction of days that were rainy, the mean rainfall over all days, and the mean rainfall over rainy days only.
#Round the means to 2 decimal places.
print('Number of rainy days:', sum(rain > 0))
print('Fraction of rainy days:',round(sum(rain > 0) / len(rain), 2))
print('Mean rainfall over all days:', round(sum(rain) / len(rain), 2))
print('Mean rainfall over rainy days:', round(sum(rain[rain > 0]) / sum(rain > 0), 2))


Number of rainy days: 8
Fraction of rainy days: 0.38
Mean rainfall over all days: 2.57
Mean rainfall over rainy days: 6.74


(b) (1 point) A town planner asks: "How much water does this area receive over time?" A hiker asks: "If it rains on my trip, how wet will I get?" Match each person to the mean that answers their question.

(c) (2 points) A report states "average daily rainfall was 2.57 mm" with no other context. Using Module 2's data - information - knowledge progression, explain why this number, by itself, is data rather than information. What context do we need to promote it to information?

**Your answer:**

(b) Town planner: Mean rainfall over all days. Hiker: Mean rainfall over rainy days

(c) We lack the context of the scope of the rainfall for the given average daily rainfall. Like the previous question (b), the average can change depending on rain over all days, or depending on if, when it is raining, how much it rains. This context can promote this data to information because now we can understand what to use the data for explaining and understanding.
